File: src/main/java/io/reactiverse/es4x/nashorn/runtime/Console.java
/*
 * Copyright 2018 Red Hat, Inc.
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  and Apache License v2.0 which accompanies this distribution.
 *
 *  The Eclipse Public License is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  The Apache License v2.0 is available at
 *  http://www.opensource.org/licenses/apache2.0.php
 *
 *  You may elect to redistribute this code under either of these licenses.
 */
package io.reactiverse.es4x.nashorn.runtime;

import io.vertx.core.Vertx;
import jdk.nashorn.api.scripting.AbstractJSObject;
import jdk.nashorn.api.scripting.JSObject;
import jdk.nashorn.api.scripting.ScriptUtils;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Console {

  private static final Pattern FORMAT = Pattern.compile("%[sdfj%]");

  private static final String RESET = "\u001B[0m";
  private static final String BOLD = "\u001B[1m";
  private static final String RED = "\u001B[31m";
  private static final String GREEN = "\u001B[32m";
  private static final String YELLOW = "\u001B[33m";
  private static final String BLUE = "\u001B[34m";

  private static String replace(String source, Pattern pattern, Function<String, String> fn) {
    final Matcher m = pattern.matcher(source);
    boolean result = m.find();
    if (result) {
      StringBuilder sb = new StringBuilder(source.length());
      int p = 0;
      do {
        sb.append(source, p, m.start());
        sb.append(fn.apply(m.group()));
        p = m.end();
      } while (m.find());
      sb.append(source, p, source.length());
      return sb.toString();
    }
    return source;
  }

  private static String format(JSObject stringify, Object... args) {
    if (!(args[0] instanceof String)) {
      List<String> objects = new ArrayList<>();
      for (Object arg : args) {
        objects.add((String) stringify.call(null, arg));
      }
      return String.join(" ", objects);
    }

    if (args.length == 1) {
      return (String) args[0];
    }

    final AtomicInteger i = new AtomicInteger(1);
    int len = args.length;
    String str = replace((String) args[0], FORMAT, x -> {
      if (x.equals("%%")) {
        return "%";
      }

      if (i.get() >= len) {
        return x;
      }

      Object v = args[i.getAndIncrement()];

      if (v == null) {
        return "null";
      }

      switch (x) {
        case "%s":
          return v.toString();
        case "%d":
          if (v instanceof Integer) {
            return Integer.toString((Integer) v);
          }
          if (v instanceof Long) {
            return Long.toString((Long) v);
          }
          break;
        case "%f":
          if (v instanceof Float) {
            return Float.toString((Float) v);
          }
          if (v instanceof Double) {
            return Double.toString((Double) v);
          }
          break;
        case "%j":
          try {
            return (String) stringify.call(null, v);
          } catch (Throwable e) {
            return "[Circular]";
          }
      }
      return x;
    });

    if (i.get() == len) {
      return str;
    }

    StringBuilder sb = new StringBuilder(str);

    do {
      Object x = args[i.get()];
      if (x == null) {
        sb.append(" null");
      } else {
        sb.append(" ");
        if (x instanceof JSObject) {
          sb.append(stringify.call(null, x));
        } else {
          sb.append(x.toString());
        }
      }
    } while (i.incrementAndGet() < len);

    return sb.toString();
  }


  private Console() {
    throw new RuntimeException("Should not be instantiated");
  }

  public static void install(Map<String, Object> bindings) {
    // get a reference to the global object
    final JSObject global = (JSObject) bindings.get("global");
    assert global != null;
    // get a reference to vertx instance
    final Vertx vertx = (Vertx) global.getMember("vertx");
    assert vertx != null;
    // get a reference to the "undefined type"
    final Object undefined = ((JSObject) global.eval("[undefined]")).getSlot(0);
    assert undefined != null;
    // we need a reference to the stringify function
    final JSObject stringify = (JSObject) ((JSObject) global.getMember("JSON")).getMember("stringify");
    assert stringify != null;

    final JSObject console = (JSObject) global.eval("new Object()");

    // state
    final Map<String, Integer> counters = new HashMap<>();
    final Map<String, Long> timers = new HashMap<>();

    console.setMember("log", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        System.out.println(format(stringify, args));
        return undefined;
      }
    });

    console.setMember("debug", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        System.out.println(GREEN + format(stringify, args) + RESET);
        return undefined;
      }
    });

    console.setMember("info", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        System.out.println(BLUE + format(stringify, args) + RESET);
        return undefined;
      }
    });

    console.setMember("warn", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        System.out.println(YELLOW + format(stringify, args) + RESET);
        return undefined;
      }
    });

    console.setMember("error", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        System.out.println(RED + format(stringify, args) + RESET);
        return undefined;
      }
    });

    console.setMember("assert", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        boolean test;

        try {
          test = (Boolean) ScriptUtils.convert(args[0], Boolean.class);
        } catch (RuntimeException e) {
          test = false;
        }

        if (test) {
          Object[] _args = new Object[args.length - 1];
          System.arraycopy(args, 1, _args, 0, _args.length);
          System.out.println(RED + format(stringify, _args) + RESET);
        }
        return undefined;
      }
    });

    console.setMember("count", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        if (args != null && args.length > 0 && args[0] != null) {
          String label = args[0].toString();

          int counter = 0;

          if (counters.containsKey(label)) {
            counter = counters.get(label);
          }

          // update
          counters.put(label, ++counter);
          System.out.println(GREEN + label + ":" + RESET + " " + counter);
        }
        return undefined;
      }
    });

    console.setMember("trace", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        if (args != null && args.length > 0 && args[0] != null) {
          if (args[0] instanceof Throwable) {
            final Throwable e = (Throwable) args[0];
            final StringWriter sw = new StringWriter();
            final PrintWriter pw = new PrintWriter(sw);
            e.printStackTrace(pw);
            final String trace = sw.toString();
            if (trace != null) {
              int idx = trace.indexOf('\n');
              if (idx != -1) {
                System.out.println(BOLD + RED + trace.substring(0, idx) + RESET + trace.substring(idx));
                return undefined;
              }
            }
          }
          System.out.println(BOLD + RED + args[0] + RESET);
        }
        return undefined;
      }
    });


    console.setMember("time", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        if (args != null && args.length > 0 && args[0] != null) {
          timers.put(args[0].toString(), System.currentTimeMillis());
        }
        return undefined;
      }
    });

    console.setMember("timeEnd", new AbstractJSObject() {
      @Override
      public boolean isFunction() {
        return true;
      }

      @Override
      public Object call(Object thiz, Object... args) {
        if (args != null && args.length > 0 && args[0] != null) {
          final long now = System.currentTimeMillis();
          String label = args[0].toString();
          if (timers.containsKey(label)) {
            System.out.println(GREEN + label + ":" + RESET + " " + (now - timers.get(label)) + "ms");
            timers.remove(label);
          } else {
            System.out.println(RED + label + ":" + RESET + " <no timer>");
          }
        }
        return undefined;
      }
    });

    bindings.put("console", console);
  }
}

File: src/main/resources/io/reactiverse/es4x/polyfill/console.js
/**
 *  Copyright 2014-2018 Red Hat, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License")
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
(function (global) {
  'use strict';

  var System = Java.type('java.lang.System');

  var formatRegExp = /%[sdj%]/g;

  function format(f) {
    if (typeof f !== 'string') {
      var objects = [];
      for (var index = 0; index < arguments.length; index++) {
        objects.push(JSON.stringify(arguments[index]));
      }
      return objects.join(' ');
    }

    if (arguments.length === 1) return f;

    var i = 1;
    var args = arguments;
    var len = args.length;
    var str = String(f).replace(formatRegExp, function(x) {
      if (x === '%%') return '%';
      if (i >= len) return x;
      switch (x) {
        case '%s': return String(args[i++]);
        case '%d': return Number(args[i++]);
        case '%j':
          try {
            return JSON.stringify(args[i++]);
          } catch (_) {
            return '[Circular]';
          }
        // falls through
        default:
          return x;
      }
    });
    for (var x = args[i]; i < len; x = args[++i]) {
      if (x === null || (typeof x !== 'object' && typeof x !== 'symbol')) {
        str += ' ' + x;
      } else {
        str += ' ' + JSON.stringify(x);
      }
    }
    return str;
  }

  var counters = {};
  var timers = {};

  var RESET = '\u001B[0m';
  var BOLD = '\u001B[1m';
  var RED = '\u001B[31m';
  var GREEN = '\u001B[32m';
  var YELLOW = '\u001B[33m';
  var BLUE = '\u001B[34m';

  global['console'] = {
    'assert': function (expression, message) {
      if (!expression) {
        System.out.println(format.apply(null, [RED + message + RESET]));
      }
    },

    count: function (label) {
      var counter;

      if (label) {
        if (counters.hasOwnProperty(label)) {
          counter = counters[label];
        } else {
          counter = 0;
        }

        // update
        counters[label] = ++counter;
        System.out.println(format.apply(null, [GREEN + label + ':' + RESET, counter]));
      }
    },

    debug: function () {
      var args = Array.prototype.slice.call(arguments);
      if (args.length > 0) {
        args[0] = GREEN + args[0];
        args[args.length - 1] = args[args.length - 1] + RESET;
      }
      System.out.println(format.apply(null, args));
    },

    info: function () {
      var args = Array.prototype.slice.call(arguments);

      if (args.length > 0) {
        args[0] = BLUE + args[0];
        args[args.length - 1] = args[args.length - 1] + RESET;
      }
      System.out.println(format.apply(null, args));
    },

    log: function () {
      System.out.println(format.apply(null, arguments));
    },

    warn: function () {
      var args = Array.prototype.slice.call(arguments);

      if (args.length > 0) {
        args[0] = YELLOW + args[0];
        args[args.length - 1] = args[args.length - 1] + RESET;
      }
      System.out.println(format.apply(null, args));
    },

    error: function () {
      var args = Array.prototype.slice.call(arguments);

      if (args.length > 0) {
        args[0] = RED + args[0];
        args[args.length - 1] = args[args.length - 1] + RESET;
      }
      System.out.println(format.apply(null, args));
    },

    trace: function (e) {
      if (e.stack) {
        // isolate first the first line
        var idx = e.stack.indexOf('\n');

        var msg = BOLD + RED + e.stack.substr(0, idx) + RESET;
        var trace = e.stack.substr(idx);

        System.out.println(format.apply(null, [msg + trace]));
      } else {
        if (e.printStackTrace) {
          e.printStackTrace(System.out);
        } else {
          System.out.println(format.apply(null, [BOLD + RED + e + RESET]));
        }
      }
    },

    time: function (label) {
      if (label) {
        timers[label] = System.currentTimeMillis();
      }
    },
    timeEnd: function (label) {
      if (label) {
        var now = System.currentTimeMillis();
        if (timers.hasOwnProperty(label)) {
          System.out.println(format.apply(null, [GREEN + label + ':' + RESET, (now - timers[label]) + 'ms']));
          delete timers[label];
        } else {
          System.out.println(format.apply(null, [RED + label + ':' + RESET, '<no timer>']));
        }
      }
    }
  };
})(this);
